Allow editing max spend limit

Gudahtt opened the pull request:
In the case where the initial spend limit for the `approve` function was set to the maximum amount, editing this value would result in the new limit being silently ignored. The transaction would be submitted with the original max spend limit.

This occurred because function to generate the new custom data would look for the expected spend limit in the existing data, then bail if it was not found (and in these cases, it was never found)

The reason the value was not found is that it was erroneously being converted to a `Number`. A JavaScript `Number` is not precise enough to represent larger spend limits, so it would give the wrong hex value (after rounding had taken place in the conversion to a floating-point number).

The data string is now updated without relying upon the original token value; the new value is inserted after the `spender` argument instead, as the remainder of the `data` string is guaranteed to be the original limit. Additionally, the conversion to a `Number` is now omitted so that the custom spend limit is encoded correctly.

Fixes #7915

Gudahtt committed:
Allow editing max spend limit

In the case where the initial spend limit for the `approve` function
was set to the maximum amount, editing this value would result in the
new limit being silently ignored. The transaction would be submitted
with the original max spend limit.

This occurred because function to generate the new custom data would
look for the expected spend limit in the existing data, then bail if
it was not found (and in these cases, it was never found).

The reason the value was not found is that it was erroneously being
converted to a `Number`. A JavaScript `Number` is not precise enough to
represent larger spend limits, so it would give the wrong hex value
(after rounding had taken place in the conversion to a floating-point
number).

The data string is now updated without relying upon the original token
value; the new value is inserted after the `spender` argument instead,
as the remainder of the `data` string is guaranteed to be the original
limit. Additionally, the conversion to a `Number` is now omitted so
that the custom spend limit is encoded correctly.

Fixes #7915

Gudahtt commented on ui/app/pages/confirm-approve/confirm-approve.util.js:
```diff
@@ -1,28 +1,27 @@
-    return customTxParamsData
+  let customPermissionValue = decimalToHex(calcTokenValue(customPermissionAmount, decimals))
+  if (customPermissionValue.length > tokenValue.length) {
+    throw new Error('Custom value too large')
```
I would like to ensure this case is caught on the edit permissions modal, so this branch is unreachable. Throwing an error here on `develop` results in the render crashing and the error page showing, but on `production` (e.g. on React v15) it silently crashes, so the user might think that it worked (though the UI doesn't update with the new limit at least, so it's not _super_ misleading)

Gudahtt replied:
Since this is a distinct and pre-existing problem, it should be fine to deal with this in a separate PR. I'm working on this now though.

Gudahtt replied:
This has been addressed in #7920

danjm commented on ui/app/pages/confirm-approve/confirm-approve.util.js:
```diff
@@ -1,28 +1,27 @@
-      return data
-    }
-    let dataWithoutCurrentAmount = matches[1]
-    const customPermissionValue = decimalToHex(calcTokenValue(Number(customPermissionAmount), decimals))
```
Is this the erroneous conversion to `Number` mentioned in the PR description?

Gudahtt replied:
Technically no, it's on line 46 of `confirm-approve.container.js`, and it's still there.

Arguably it's no longer a bug if it's only used for display? 🤔 So I hadn't removed it.

There was a second case of erroneously converting to `Number` as well, which is here.

Gudahtt replied:
Actually maybe it is a bug - I'll revisit this, to prevent rounding errors in the displayed value.

danjm replied:
Right. I understand now. The match that was failing was on line 9.

Gudahtt replied:
Yes, exactly

Gudahtt replied:
I've removed the problematic line, and am now passing the token limits and amount as strings: https://github.com/MetaMask/metamask-extension/pull/7919/commits/fb39ad75c40a1ec09861245b0c3cf5688e9011e0

danjm commented on ui/app/pages/confirm-approve/confirm-approve.util.js:
```diff
@@ -1,28 +1,27 @@
-    const customTxParamsData = dataWithoutCurrentAmount + customPermissionValue
-    return customTxParamsData
+  let customPermissionValue = decimalToHex(calcTokenValue(customPermissionAmount, decimals))
+  if (customPermissionValue.length > tokenValue.length) {
```
Why `customPermissionValue.length > tokenValue.length`?

`tokenValue` is the approve amount set by the dapp correct? If `tokenValue` < `customPermissionValue` < [largest possible token value], wouldn't this exclude some valid `customPermissionValue`s?

danjm replied:
Oh, wait, I get it

danjm replied:
Actually, I don't. Why can't we have a `customPermissionValue` of length greater than `tokenValue` length?

Gudahtt replied:
`tokenValue` here is the zero-padded token value, which is always of length 64 (which is how much space a `u256` takes up in hex). I should probably explain this in the code - and maybe assert my expectations then check for greater-than-64 explicitly 🤔

danjm replied:
> and maybe assert my expectations then check for greater-than-64 explicitly
yes, and if so, we can do that validation earlier in the code

Gudahtt replied:
I'm not entirely sure what you mean; where do you imagine this validation should take place? I've updated this - let me know what you think

danjm commented on ui/app/pages/confirm-approve/confirm-approve.util.js:
```diff
@@ -1,28 +1,27 @@
+  if (!tokenData) {
+    throw new Error(`Invalid data`)
+  } else if (tokenData.name !== 'approve') {
+    throw new Error(`Invalid data; should have be 'approve' method, but instead is '${tokenData.name}'`)
```
Minor, typo in `Invalid data; should have be 'approve' method, but instead is `

Gudahtt committed:
Ensure token balance and limit are passed as strings

To avoid rounding errors, all token limts and the token balance are
passed as aribtrary precision strings. The balance is now explicitly
converted into 9-decimal precision `Number` for display, and the spend
limits are converted to `Number`s for display. This matches how they
were displayed previously.

Gudahtt committed:
Improve error handling

danjm reviewed (approved):
LGTM
